dask/delayed.py

```diff
@@ -230,7 +233,9 @@ def tokenize(*args, **kwargs):
 
 
 @curry
-def delayed(obj, name=None, pure=None, nout=None, traverse=True):
+def delayed(
+    obj: D, name=None, pure: bool = None, nout: int = None, traverse: bool = True
```

Comment:
I'm quite sure mypy/pyright require you to bind `D`: you need to use it in two places (generic return value!) or just use `Any` for `obj` if a generic return value is too complicated for now.

If the dask CI doesn't validate type annotations, you should at least locally test whether

`reveal_type(delayed(some_object))`

returns  the appropriate type when running mypy/pyright.

Reply:
Looking at the `Delayed` class https://github.com/dask/dask/blob/7122a23e04032152ce4a2c90cd3c156197473e4f/dask/delayed.py#L485

I honestly don't know which of its attributes/functions should use a generic type.

Without generics, the type annotations of `delayed` should probably be:

```py
from typing import Any, Hashable, Optional

def delayed(obj: Any, name: Hashable = None, pure: Optional[bool] = None, nout: Optional[int] = None, traverse: bool = True) -> Delayed:
    ...
```

Reply:
@tebesfinwo That is awesome.  Should there a Union for the name attribute?:

```python
from typing import Any, Hashable, Optional, Union

def delayed(obj: Any, name: Union[str, Hashable] = None, pure: Optional[bool] = None, nout: Optional[int] = None, traverse: bool = True) -> Delayed:
```

Reply:
I tried `Union`, but pytest does not like it.

Reply:
I don't think it is needed (both `str` and `None` are hashable) but it might be more readable to use `Optional[Union[str, Hashable]]`

```py
"__hash__" in dir("") # True
"__hash__" in dir(None) # True
```

Reply:
@twoertwein thank you for your help!! This PR all you!  I really like the `Any`.  I need to add this to my repertoire. 

I am going to commit another change. The pytest errors I get belie the testing the function call manually. I pulled the `-> Delayed:` which passes, while no version of `Union` seems to make it through pytest.

Reply:
str is a subclass of Hashable